lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/LengthGoalBreakIterator.java

```diff
@@ -174,7 +175,48 @@ private int moveToBreak(int idx) { // precondition: idx is a known break
   // called at start of new Passage given first word start offset
   @Override
   public int preceding(int offset) {
-    return baseIter.preceding(offset); // no change needed
+    final int fragmentStart = Math.max(baseIter.preceding(offset), 0); // convert DONE to 0
+    fragmentEndFromPreceding = baseIter.following(fragmentStart);
```

Comment:
couldn't fragmentStart argument be offset so that baseIter needn't re-analyze the text before offset?

Reply:
A point I want to make is that BreakIterators can be rather expensive and so it's good to try to avoid calling it more than needed, and to help it do less work.

Reply:
Unfortunately no. The fragmentStart argument is the start of the match that could be anything depending on the tokenizer in the index analyzer chain. Even if we assume it's the start of a word or a phrase, the underlying BI can break on different places. In case of SENTENCE the preceding() call here will find the beginning of the sentence. In case of SEPARATOR, which is customizable by query, the breaks can be anywhere else.
We could only assume fragmentStart is a break point if the underlying BI would be the same as the tokenizer in the index analyzer chain. (I'm not sure, but the query analyzer chain could be different I think.)

Reply:
A little correction: the **offset** parameter is the start of the match and the fragmentStart could be different because of the issues described above.

Reply:
I understand what `offset` param is and what `fragmentStart` is, and that the boundaries of each are in different universes that may or may not align.  But I don't see why that matters here.  I claim the call to `baseIter.following(fragmentStart)` is guaranteed to be equal to `baseIter.following(offset)` _because_ fragmentStart is `baseIter.preceding(offset)`.  The argument to the breakIterator can be any arbitrary character offset; it needn't be an existing break of any kind.  FWIW I made the simple correction and tests pass.  If you insist there is a problem then can you present a failing test?

Reply:
AH I see! I somehow misinterpreted your question. You're mostly correct. The `baseIter.following(fragmentStart)` should be `baseIter.following(offset - 1)`. The minus one is needed for the very unlikely edge case where the match is a single character and the user requests the absolute minimal length for highlight. I added a little test that checks for that, you can see for yourself.